File: testdata/fixtures/ruby/CWE-434.rb
# CWE-434: params[:file] -> File.open without allowlist
class UploadController < ApplicationController
  def vuln1
    f = params[:file]
    # ruleid: ruby-434-upload
    File.open("/uploads/" + f.original_filename, "wb") { |io| io.write(f.read) }
  end

  def vuln2
    name = params[:file].original_filename
    # ruleid: ruby-434-upload
    File.write("/tmp/" + name, params[:file].read)
  end

  def vuln3
    uploaded = params["file"]
    # ruleid: ruby-434-upload
    IO.write("/uploads/" + uploaded[:filename], uploaded.read)
  end

  def safe_allowlist
    f = params[:file]
    ext = File.extname(f.original_filename)
    if ![".png", ".jpg"].include?(ext)
      raise "bad"
    end
    # ok: ruby-434-upload
    File.open("/uploads/allowed.png", "wb") { |io| io.write(f.read) }
  end

  def safe_fixed
    # ok: ruby-434-upload
    File.open("/tmp/fixed.png", "wb") { |io| io.write("fixed") }
  end
end

File: testdata/fixtures/csharp/CWE-434.cs
// CWE-434: HttpPostedFileBase without extension allowlist
using System.IO;
using System.Web;

public class TestCWE434 : System.Web.Mvc.Controller {
    public void vuln1(HttpPostedFileBase file) {
        string name = file.FileName;
        // ruleid: csharp-434-upload
        File.SaveAs("/uploads/" + name);
    }
    public void vuln2() {
        HttpPostedFileBase f = Request.Files[0];
        string p = f.FileName;
        // ruleid: csharp-434-upload
        File.WriteAllBytes("/tmp/" + p, new byte[0]);
    }
    public void vuln3(HttpPostedFileBase uploaded) {
        var path = uploaded.FileName;
        // ruleid: csharp-434-upload
        File.Create("/uploads/" + path);
    }
    public void safeAllowlist(HttpPostedFileBase file) {
        string ext = Path.GetExtension(file.FileName);
        if (ext != ".png" && ext != ".jpg") return;
        string allowlist = "/uploads/allowed.png";
        // ok: csharp-434-upload
        File.SaveAs(allowlist);
    }
    public void safeFixed() {
        // ok: csharp-434-upload
        File.WriteAllBytes("/tmp/fixed.png", new byte[0]);
    }
}
